src/ray/gcs/gcs_server/gcs_table_storage.h

```diff
@@ -0,0 +1,382 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RAY_GCS_GCS_TABLE_STORAGE_H_
+#define RAY_GCS_GCS_TABLE_STORAGE_H_
+
+#include <utility>
+
+#include "ray/gcs/store_client/redis_store_client.h"
+#include "ray/protobuf/gcs.pb.h"
+
+namespace ray {
+namespace gcs {
+
+using rpc::ActorCheckpointData;
+using rpc::ActorCheckpointIdData;
+using rpc::ActorTableData;
+using rpc::ErrorTableData;
+using rpc::GcsNodeInfo;
+using rpc::HeartbeatBatchTableData;
+using rpc::HeartbeatTableData;
+using rpc::JobTableData;
+using rpc::ObjectTableData;
+using rpc::ObjectTableDataList;
+using rpc::ProfileTableData;
+using rpc::ResourceMap;
+using rpc::ResourceTableData;
+using rpc::TaskLeaseData;
+using rpc::TaskReconstructionData;
+using rpc::TaskTableData;
+using rpc::WorkerFailureData;
+
+/// \class GcsTable
+///
+/// GcsTable supports putting, getting and deleting of GCS table data.
+/// This class is not meant to be used directly. All gcs table classes without job id
+/// should derive from this class and override the table_name_ member with a unique value
+/// for that table.
+template <typename Key, typename Data>
+class GcsTable {
+ public:
+  explicit GcsTable(std::shared_ptr<StoreClient<Key, Data, JobID>> store_client)
+      : store_client_(store_client) {}
+
+  virtual ~GcsTable() { store_client_.reset(); }
+
+  Status Put(const Key &key, const Data &value, const StatusCallback &callback);
+
+  Status Get(const Key &key, const OptionalItemCallback<Data> &callback);
+
+  Status GetAll(const SegmentedCallback<std::pair<Key, Data>> &callback);
+
+  Status Delete(const Key &key, const StatusCallback &callback);
+
+  std::string table_name_;
+  std::shared_ptr<StoreClient<Key, Data, JobID>> store_client_;
+};
+
+/// \class GcsTableWithIndex
+///
+/// GcsTableWithIndex supports putting, getting and deleting of GCS table data.
+/// This class is not meant to be used directly. All gcs table classes with job id should
+/// derive from this class and override the table_name_ member with a unique value
+/// for that table.
+template <typename Key, typename Data>
+class GcsTableWithIndex : public GcsTable<Key, Data> {
+ public:
+  explicit GcsTableWithIndex(std::shared_ptr<StoreClient<Key, Data, JobID>> store_client)
+      : GcsTable<Key, Data>(store_client) {}
+
+  Status PutWithJobId(const JobID &job_id, const Key &key, const Data &value,
```

Comment:
Shouldn't expose both `Put` and `PutWithIndex`. 
I think we can just use `Put(key, value, callback)` in `GcsTableWithIndex`, because we can get the job id from the key.

Reply:
fixed